### PR TITLE
Parser: Make parser mostly reentrant

### PR DESCRIPTION
**Pull request by bstoeger**
### Describe the pull request:

- [ ] Bug fix
- [ ] Functional change
- [ ] New feature
- [x] Code cleanup
- [ ] Build system change
- [ ] Documentation change
- [ ] Language translation

### Pull request long description:

Before doing the dive-site parsing changes [replace UUIDs by pointers] it seemed appropriate to isolate the whole global parser state. After this change-set the parser still accesses a few global variables [such as autogroup], but that should be variables that are accessed from other parts of the code.

This was mostly done "peephole" style. But it seems to read my own log and runs the tests.

There is one potentially debatable thing noted in the commit-message:
> A somewhat controversial point might be that the individual parsing functions are split in those that need parser-state and those that don't. This means that there are now two versions of the MATCH macro, viz. one for the former and one for the latter.

### Changes made:

1) Make two free_*() functions NULL aware.
2) Introduce a parser_state structure.

### Release note:

No.
### Documentation change:

No.

### CODE REVIEW

**Comment by bstoeger** [comment]
For documentation purposes, a list of static variables in the parser after this PR:

1) In `parse_float()`, `first_time` is used to emit a single warning if floating point has a comma. A comment suggests that this could be replaced with a test for subsurface files (which is now available in the parser_state structure).
2) In `add_dive_site()`, `suffix` is used to create dive site numbers. This is questionable anyway [you want a variable per dive site name, not a global one], but let's have a closer look when refactoring the dive site code.
3) In `cobalt_location()` we remember the string of the location. This should probably be moved to `parser_state`.

File: core/parse.h [hunk 1]
```diff
@@ -31,67 +20,95 @@ struct parser_settings {
...
+	int lastcylinderindex, next_o2_sensor;
+	int o2pressure_sensor;
+	struct extra_data cur_extra_data;
+	struct units xml_parsing_units;
+	struct dive_table *target_table;	/* non-owning */
+
+	sqlite3 *sql_handle;			/* for SQL based parsers */
+	event_allocation_t event_allocation;
+};
+#define cur_event event_allocation.event
```
Comment (neolit123) [inline core/parse.h:66]: add empty line above this line.

File: core/import-cobalt.c [hunk 2]
```diff
@@ -73,20 +73,19 @@ extern int cobalt_buddies(void *handle, int columns, char **data, char **column)
...
 
-extern int cobalt_location(void *handle, int columns, char **data, char **column)
+static int cobalt_location(void *param, int columns, char **data, char **column)
 {
-	UNUSED(handle);
 	UNUSED(columns);
 	UNUSED(column);
+	struct parser_state *state = (struct parser_state *)param;
 
 	static char *location = NULL;
```
Comment (neolit123) [inline core/import-cobalt.c]: i guess it makes sense to be moved in the state as a generic "location" or possibly "cobalt_location" param.

File: core/parse.h [hunk 3]
```diff
@@ -31,67 +20,94 @@ struct parser_settings {
...
+void sample_start(struct parser_state *state);
+void sample_end(struct parser_state *state);
+void divecomputer_start(struct parser_state *state);
+void divecomputer_end(struct parser_state *state);
+void userid_start(struct parser_state *state);
+void userid_stop(struct parser_state *state);
 void utf8_string(char *buffer, void *_res);
 
-void add_dive_site(char *ds_name, struct dive *dive);
+void add_dive_site(char *ds_name, struct dive *dive, struct parser_state *state);
```
Comment (neolit123) [inline core/parse.h:110]: +1 for moving suffix to state - e.g. dive_site_suffix.

**Review by neolit123 — APPROVED** [review]
thanks for the refactor. added some minor comments.
the global state of the parser has not caused us serious problems, but this seems like a good RFC.

also +1 for removing static variables.
i have not examined if CI passes.

File: core/parse.h (hunk 3 above)
Comment (bstoeger) [reply core/parse.h:110]: Stupid question: Is this `SSRF_WS` thing the webservice that was recently removed or is this something completely different? I.e. could we simply remove the code?

Comment (bstoeger) [reply core/parse.h]: Stupid question: Is this `SSRF_WS` thing the webservice that was recently removed or is this something completely different? I.e. could we simply remove the code?

Comment (bstoeger) [reply core/parse.h]: Stupid question: Is this `SSRF_WS` thing the webservice that was recently removed or is this something completely different? I.e. could we simply remove the code?

Comment (bstoeger) [reply core/parse.h]: Stupid question: Is this `SSRF_WS` thing the webservice that was recently removed or is this something completely different? I.e. could we simply remove the code?

Comment (neolit123) [reply core/parse.h:110]: ^ seems like so, but would defer to @dirkhh on the decision.
(github is spazzing today i see comments repeated 6 times)

**Comment by neolit123** [comment]
> Stupid question: Is this SSRF_WS thing the webservice that was recently removed or is this something completely different? I.e. could we simply remove the code?

seems like so, but would defer to @dirkhh on the decision.
(github is spazzing today i see comments repeated 6 times)

File: core/parse.h (hunk 3 above)
Comment (bstoeger) [reply core/parse.h]: Stupid question: Is the whole `SSRF_WS` business related to the web-service that was removed recently? I.e. could the whole code simply be deleted?

Comment (bstoeger) [reply core/parse.h:110]: Argh.. Sorry. I tried again and again and it would not display anything for me. Perhaps my browser.

**Review by bstoeger — COMMENTED** [review]
I think the problem was that I inadvertenly started a review. Trying to close that now.. :/

File: core/parse.h (hunk 3 above)
Comment (dirkhh) [reply core/parse.h:110]: no, GitHub was unwell - I saw the emails about you commenting, but couldn't see the comments on GitHub. Maybe some of their distributed infrastructure had outdated caches or something.
Anyway, yes, that SSRF_WS was indeed about the soon-to-be-discontinued GPS web service. I merged the commit removing it.

File: core/divelist.c [hunk 4]
```diff
@@ -765,9 +765,11 @@ void insert_trip_dont_merge(dive_trip_t *dive_trip)
 /* free resources associated with a trip structure */
 void free_trip(dive_trip_t *trip)
 {
-	free(trip->location);
-	free(trip->notes);
-	free(trip);
+	if (trip) {
```
[the remark this answers is not in the document]
Comment (bstoeger) [reply core/divelist.c:768]: Yes. But I tried to model after the other `free_*()` functions, which seem to prefer the 
```
	if (x) {
		free(...)
	}
```
scheme. I would convert them all in one go, if this is desired.

File: core/import-cobalt.c (hunk 2 above)
Comment (bstoeger) [reply core/import-cobalt.c]: I added a second commit, which makes the code more explicit. But this is only compile tested! Do we have Cobalt files? I don't see any obvious things in the dives-directory.

**Comment by bstoeger** [comment]
Replying via e-mail since for some reason I can't post on github!?

> -void add_dive_site(char *ds_name, struct dive *dive);
> +void add_dive_site(char *ds_name, struct dive *dive, struct parser_state
> *state);
> 
> +1 for moving suffix to state - e.g. dive_site_suffix.

Stupid question: Is the whole `SSRF_WS` business related to the web-service 
that was removed recently? I.e. could the whole code simply be deleted?

**Comment by dirkhh** [comment]
Typo in the last commit message: on vs. one
I'm still reading and re-reading the big commit. And with travel and jet lag and now conference I haven't quite found enough time to do a credible review. Sorry. Hopefully later today.

File: core/parse.h [hunk 5]
```diff
@@ -31,67 +20,95 @@ struct parser_settings {
...
+};
+
+/*
+ * parser_state is the state needed by the parser(s). It is initialized
+ * with init_parser_state() and resources are freed with free_parser_state().
+ * "owning" marks pointers to objects that are freed in free_parser_state().
+ * In contrast, "non-owning" marks pointers to objects that are owned
+ * by other data-structures.
+ */
+struct parser_state {
```
[the remark this answers is not in the document]
Comment (dirkhh) [reply core/parse.h:38]: there is no rule. or I should say "there no longer is"
Linus initially had some ideas that having a XXX_t meant that this was a type by itself vs a convenient collection of variables (or something like this - I'd have to go through my email archive for more detail), but I think this has long long gone out the window.

**Review by dirkhh — APPROVED** [review]
Somehow reading this a third time it suddenly all made sense. No idea why I was stuck the first two times I tried.
Great effort. I hate huge commits like this, but this was the only way to do it, I think.
I agree with the MATCH/MATCH_STATE split.

**Comment by dirkhh** [comment]
last commit message slightly edited and all of it merged.

**Pull request closed without merging**